table/block_based_table_reader.cc

```diff
@@ -2277,18 +2286,27 @@ InternalIterator* BlockBasedTable::NewRangeTombstoneIterator(
     Cache* block_cache = rep_->table_options.block_cache.get();
     assert(block_cache != nullptr);
     if (block_cache->Ref(rep_->range_del_entry.cache_handle)) {
-      auto iter = rep_->range_del_entry.value->NewIterator<DataBlockIter>(
-          &rep_->internal_comparator,
-          rep_->internal_comparator.user_comparator());
+      iter = std::unique_ptr<InternalIterator>(
+          rep_->range_del_entry.value->NewIterator<DataBlockIter>(
+              &rep_->internal_comparator,
+              rep_->internal_comparator.user_comparator()));
       iter->RegisterCleanup(&ReleaseCachedEntry, block_cache,
                             rep_->range_del_entry.cache_handle);
-      return iter;
     }
-  }
-  // The meta-block exists but isn't in uncompressed block cache (maybe
-  // because it is disabled), so go through the full lookup process.
-  return NewDataBlockIterator<DataBlockIter>(rep_, read_options,
-                                             rep_->range_del_handle);
+  } else {
+    // The meta-block exists but isn't in uncompressed block cache (maybe
+    // because it is disabled), so go through the full lookup process.
+    iter = std::unique_ptr<InternalIterator>(
+      NewDataBlockIterator<DataBlockIter>(
+          rep_, read_options, rep_->range_del_handle));
+  }
+  auto new_tombstone_fragments = std::make_shared<FragmentedRangeTombstoneList>(
```

Comment:
There's a race here. If two threads try to call `NewRangeTombstoneIterator`, they may both create a new `FragmentedRangeTombstoneList` and overwrite `rep_->fragmented_tombstones`. Probably need to serialize access here.